### PR TITLE
security/p2p: prevent peers who errored being added to the peer_set

### PR DESCRIPTION
**Pull request by jmalicevic**
This work is a fix for a bug in the P2P layer.

A node can be attacked via the p2p layer by saturating its incoming connection slots and not allowing the node to accept new conditions. This happens when an attacker continuously submits requests to connect with an erroneous message causing the incoming request to error before it has been accepted. The attacked node, tries to remove the peer from its peer set which silently fails (due to the peer not yet being in the peer set). The routine adding a peer into the peer set happens in parallel in the background and will add the peer after the error has been reported. 

This fix resolves the issue in the following way:

1. We add a field  `removalAttemptFailed` to the `Peer` datastructure. 
2. If removal of this peer fails, we set it to true. 
3. When adding a peer into the peer set, the `Add` function will return an `ErrPeerRemoval` error if this field was true and *not* add the peer.  

**Note**. This attack does not work if the config flag `allow_duplicate_ips` is set to `false`. 

---

#### PR checklist

- [ ] Tests written/updated, or no tests needed
- [ ] `CHANGELOG_PENDING.md` updated, or no changelog entry needed
- [ ] Updated relevant documentation (`docs/`) and code comments, or no
      documentation updates needed

### CODE REVIEW

File: p2p/switch.go [hunk 1]
```diff
@@ -370,6 +370,12 @@ func (sw *Switch) stopAndRemovePeer(peer Peer, reason interface{}) {
 	// https://github.com/tendermint/tendermint/issues/3338
 	if sw.peers.Remove(peer) {
 		sw.metrics.Peers.Add(float64(-1))
+	} else {
+		// Removing the peer has failed so we set a flag to mark that a removal was attempted
+		// This is done to avoid the corner case of adding a peer when this error happens before the
+		// AddPeer routine has finished
```
Comment (cmwaters) [inline p2p/switch.go:376]: Can you just extend this to inform the reader what setting that flag means i.e. that once it's set the node won't add that peer again

File: p2p/switch.go [hunk 2]
```diff
@@ -370,6 +370,12 @@ func (sw *Switch) stopAndRemovePeer(peer Peer, reason interface{}) {
...
+		peer.SetRemovalFailed()
+		sw.Logger.Info("error on peer removal", ",", "peer", peer.ID())
```
Comment (cmwaters) [inline p2p/switch.go:378]:
```suggestion
		sw.Logger.Debug("error on peer removal", ",", "peer", peer.ID())
```
Do we expect a node operator to do anything when this happens. If not, I'd suggest lowering it to Debug

File: p2p/peer_set.go [hunk 3]
```diff
@@ -47,6 +47,9 @@ func (ps *PeerSet) Add(peer Peer) error {
 	if ps.lookup[peer.ID()] != nil {
 		return ErrSwitchDuplicatePeerID{peer.ID()}
 	}
+	if peer.GetRemovalFailed() {
+		return ErrPeerRemoval{}
+	}
```
Comment (cmwaters) [inline p2p/peer_set.go:52]: Can we do this check before calling `PeerSet.Add` here. I would prefer to do it on line 813 of switch.go before initializing the peer to the individual reactors

Comment (thanethomson) [reply p2p/peer_set.go:52]: Will this not change the effectiveness of the fix?

Comment (cmwaters) [reply p2p/peer_set.go:52]: What do you mean by effectiveness? As in it will no longer fix it?

Comment (jmalicevic) [reply p2p/peer_set.go:52]: If this check happens anywhere earlier we introduce the exact same race condition we already heaving.

File: p2p/switch.go (hunk 1 above)
Comment (jmalicevic) [reply p2p/switch.go]: Done.

File: p2p/peer_set.go (hunk 3 above)
Comment (cmwaters) [reply p2p/peer_set.go:52]: > If this check happens anywhere earlier we introduce the exact same race condition we already heaving.

I see. I wasn't aware of that. Need to look in more closely then

File: p2p/switch.go [hunk 4]
```diff
@@ -370,6 +370,13 @@ func (sw *Switch) stopAndRemovePeer(peer Peer, reason interface{}) {
 	// https://github.com/tendermint/tendermint/issues/3338
 	if sw.peers.Remove(peer) {
 		sw.metrics.Peers.Add(float64(-1))
+	} else {
+		// Removing the peer has failed so we set a flag to mark that a removal was attempted
+		// This is done to avoid the corner case of adding a peer when this error happens before the
+		// AddPeer routine has finished. Setting this flag will prevent a peer from being added
+		// to a node's peer set.
+		peer.SetRemovalFailed()
```
Comment (thanethomson) [inline p2p/switch.go:378]: Is it safe to do this here without the peer set being locked in the same way it is when calling `GetRemovalFailed` [here](https://github.com/tendermint/tendermint/pull/9500/files#diff-9017d44e6c3e27f96079fbe86494298ad63faef9c9c615fe2f499dd5a3765c34R50)? (the `GetRemovalFailed` call happens while the peer set's mutex is locked)

Comment (jmalicevic) [reply p2p/switch.go:378]: I had intuitively implemented with a lock in a previous version (although by setting a lock on the peer itself, not on the entire set). But I did not notice the attack being more successful so I opted to not add a mutex into the peer structure itself even though I still do a small window of opportunity for the removal to happen first and then the Add to grab a lock and add the peer before we set the removal failed flag. Thinking about it now, I could actually set the flag inside the Remove function above the line that returns [false](https://github.com/tendermint/tendermint/blob/499804c6232f372a2ff1d753889f7d5c8f8b33c3/p2p/peer_set.go#L113) and not do a lock on the peer itself. I will do a local test with this change to examine the behaviour unless you have objections.

Comment (jmalicevic) [reply p2p/switch.go:378]: Based on a local test I think it is better to put it inside the Remove. Will confirm on the testnet before pushing the change.

**Review by thanethomson — APPROVED** [review]
I'd say we should go ahead and merge this. Next steps from here:
- [ ] Backport to `v0.37.x` and `v0.34.x`, and add a `CHANGELOG_PENDING.md` entry to each backport PR.
- [ ] Run a large-scale testnet (a single test) on `v0.34.x` with this fix and check that things work as expected.
- [ ] Cut v0.34.22.
- [ ] Keep the `v0.37.x` backport branch open until we've cut v0.37.0, then merge it.
- [ ] Run another large-scale testnet on `v0.37.x` to check that things work before cutting v0.37.1.

Unless anyone has any objections there?

**Comment by jmalicevic** [comment]
> 

Thanks! I'll create an issue with the items you listed, I think they all make sense are definitely worth doing. Do you think we should do a large testnet test of whether the attacked is prevented? If so, we can add this item to the list and I'll need to patch the testnet scripts to be able to execute the attack.

**Comment by thanethomson** [comment]
> Do you think we should do a large testnet test of whether the attacked is prevented? If so, we can add this item to the list and I'll need to patch the testnet scripts to be able to execute the attack.

The main thing we'd need to know is whether the P2P network is still stable and that the fix doesn't meaningfully affect peer connectivity. I don't think it needs to be a long test. I assume that your previous tests showed that the fix worked in larger-scale testnets?

**Comment by jmalicevic** [comment]
> I assume that your previous tests showed that the fix worked in larger-scale testnets?

I tested that a more aggressive attack does not bring the node down and that the attack is prevented. I did not run a 200node test with the attack. I also don;t think it would reveal additional information. But it was worth testing that when we multiple the attackers  the node is still behaving correctly. So I agree with you, we should confirm that this fix does not break something else rather then attack a node within a bigger network.

**Comment by cason** [comment]
This is an elegant workaround. As Jasmina mentioned, while I agree we should assess the p2p layer as a whole in large testnet, I don't see how a larger environment would affect this form of attack or compromise the solution.

**Review by 09125631619 — APPROVED** [review]

**Pull request merged**